### PR TITLE
Make underscore_literal_suffix a hard error.

### LINKED ISSUE #42326: `underscore_literal_suffix` future-compatibility warnings
This is the **summary issue** for the `UNDERSCORE_LITERAL_SUFFIX` future-compatibility warning and other related errors. The goal of this page is describe why this change was made and how you can fix
code that is affected by it. It also provides a place to ask questions or register a complaint if you feel the change should not be made.

#### What is the warning for?

The `underscore_literal_suffix` warning is issued when you use `_` as a literal suffix. Due to the behavior of the lexer, this is limited to these two situations:

- After char/string/byte/bytestring literals: `"Foo"_`
- After float literals ending with `.`: `1._` (this kind of construction is already banned by means of #41946)

This warning can always be fixed by removing the `_` and inserting whitespaces if necessary.

#### When will this warning become a hard error?
                    
At the beginning of each 6-week release cycle, the Rust compiler team
will review the set of outstanding future compatibility warnings and
nominate some of them for **Final Comment Period**. Toward the end of
the cycle, we will review any comments and make a final determination
whether to convert the warning into a hard error or remove it
entirely.

### PR DESCRIPTION
**Pull request by nnethercote**
It's been a warning for 5.5 years. Time to make it a hard error.

Closes #42326.

r? @pnkfelix

### CODE REVIEW

**Comment by nnethercote** [comment]
From [#42326](https://github.com/rust-lang/rust/issues/42326#issuecomment-1072498296):

> Next steps are:
>
> 1. Run crater on experiementally moving it to hard error, to evaluate fallout.
> 2. If fallout is non-trivial, then it needs to be moved to a proper lint based implementation
> 3. Likewise, if fallout is non-trival, then ensure that it is included in the cargo report future-incompat output, when relevant.
> 4. If fallout is trivial, then we can just move to hard error without going through the longer route of replacing with a lint and whatnot.

So let's do a try run, which is a prereq for a crater run.
@bors try

**Comment by nnethercote** [comment]
@craterbot run mode=check-only

**Comment by nnethercote** [comment]
I don't know how to read crater reports. But from clicking around randomly I found at least one clear regression:
```
[INFO] [stderr]     Checking simple-replace-templete-engine v0.1.0 (/opt/rustwide/workdir)
[INFO] [stdout] error: underscore literal suffix is not allowed
[INFO] [stdout]   --> src/lib.rs:83:61
[INFO] [stdout]    |
[INFO] [stdout] 83 |         let templete_str = "hello _t_name_t_ _t_sec_name_t_"_.to_owned();
[INFO] [stdout]    |                                                    
```
But I can't find that crate on crates.io or anywhere else.

**Comment by Nilstrieb** [comment]
If you click on the name you'll get sent to the Github repository of the crate. Zero stars and the last commit from 3 years. The author is still active on Github, so I'll just send in a PR.

**Comment by nnethercote** [comment]
From above:

> 4. If fallout is trivial, then we can just move to hard error without going through the longer route of replacing with a lint and whatnot.

AFAICT we have a single failure in the crater run, for a crate that hasn't been touched in 3 years and has few if any users. @Nilstrieb has kindly filed a [PR](https://github.com/woodgear/simple-replace-templete-engine) with the trivial fix anyway.

So I think we're good to land this.

r? @petrochenkov

File: src/test/ui/parser/underscore-suffix-for-string.rs [hunk 1]
```diff
@@ -1,8 +1,19 @@
-// check-pass
+// check-fail
```
Comment (petrochenkov) [inline src/test/ui/parser/underscore-suffix-for-string.rs:1]: `check-fail` is the default, no need to specify it explicitly.

**Comment by petrochenkov** [comment]
r=me after cleaning up the test.
@rustbot author

**Pull request merged**